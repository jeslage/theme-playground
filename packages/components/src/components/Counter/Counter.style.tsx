import styled from 'styled-components';

const StyledCounter = styled.div`
  display: flex;
  align-items: center;
  width: 100%;

  .counter__counter {
    display: flex;
    align-items: center;
  }

  button {
    padding: 0;
    margin: 0;
    outline: none;
    background: none;
    border: 1px solid ${({ theme }) => theme.text};
    width: 30px;
    height: 30px;
    border-radius: 15px;
    display: inline-flex;
    align-items: center;
    justify-content: center;
    cursor: pointer;

    &[disabled] {
      opacity: 0.2;
      pointer-events: none;
    }

    svg {
      fill: ${({ theme }) => theme.text};
      width: 22px;
      height: 22px;
    }
  }

  span {
    margin: 0 5px;
  }

  input {
    border: none;
    border-bottom: 1px solid transparent;
    outline: none;
    display: inline-block;
    max-width: 60px;
    font-size: 14px;
    padding: 5px 10px 4px;
    font-family: inherit;
    text-align: center;
    color: inherit;
    background: none;

    &:focus {
      border-color: ${({ theme }) => theme.text};
    }
  }
`;

export default StyledCounter;
