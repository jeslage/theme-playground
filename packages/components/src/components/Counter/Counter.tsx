import * as React from 'react';
import { HandleChange } from '../../definitions';

import StyledCounter from './Counter.style';
import Label from '../Label/Label';

export interface Props {
  iconBefore?: HTMLElement;
  label: string;
  onChange: (val: number, suffix: string | undefined) => void;
  value: number;
  description?: string;
  min?: number;
  max?: number;
  steps?: number;
  suffix?: string | undefined;
}

const countDecimals = (number: number) => {
  if (Math.floor(number) === number) return 0;
  return number.toString().split('.')[1].length || 0;
};

const Counter: React.FC<Props> = ({
  label,
  description,
  min = 0,
  max = 100,
  steps = 1,
  onChange,
  value,
  suffix,
  iconBefore
}) => {
  const [counterValue, setCounterValue]: any = React.useState(value);

  const fixedNumber = countDecimals(steps);
  const updateValue = (val: number) => {
    setCounterValue(val);
    onChange(val, suffix);
  };

  const handleChange = (event: HandleChange) => {
    const { value: eventValue, validity } = event.target;

    if (validity.valid) {
      setCounterValue(eventValue);
    }
  };

  const handleBlur = (event: HandleChange) => {
    const { value: eventValue } = event.target;
    const numberValue: number | string = parseFloat(eventValue);

    if (numberValue > max) {
      updateValue(max);
    } else if (numberValue < min || Number.isNaN(numberValue)) {
      updateValue(min);
    } else {
      updateValue(numberValue);
    }
  };

  return (
    <StyledCounter>
      <Label iconBefore={iconBefore} label={label} description={description} />

      <div className="counter__counter">
        <button
          type="button"
          onClick={() =>
            updateValue(
              counterValue - steps >= min
                ? parseFloat((counterValue - steps).toFixed(fixedNumber))
                : min
            )
          }
          disabled={counterValue === min}
          aria-label="Decrease"
        >
          <svg viewBox="0 0 24 24" width="24" height="24">
            <rect height="2" rx="1" width="12" x="6" y="11" />
          </svg>
        </button>
        <span>
          <input
            type="text"
            pattern="[0-9.]*"
            value={counterValue}
            onChange={handleChange}
            onBlur={handleBlur}
          />
          {suffix}
        </span>
        <button
          type="button"
          onClick={() =>
            updateValue(
              counterValue + steps <= max
                ? parseFloat((counterValue + steps).toFixed(fixedNumber))
                : max
            )
          }
          disabled={counterValue === max}
          aria-label="Increase"
        >
          <svg viewBox="0 0 24 24" width="24" height="24">
            <rect height="2" rx="1" width="12" x="6" y="11" />
            <rect height="12" rx="1" width="2" x="11" y="6" />
          </svg>
        </button>
      </div>
    </StyledCounter>
  );
};

export default Counter;
