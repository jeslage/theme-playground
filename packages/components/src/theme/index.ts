export const darkTheme = {
  base: 'dark',
  backgroundDark: '#2f2f2f',
  backgroundLight: '#333',
  border: '#999999',
  text: '#FFFFFF'
};

export const lightTheme = {
  base: 'light',
  backgroundDark: '#f6f9fc',
  backgroundLight: '#fff',
  border: '#666666',
  text: '#333333'
};

export const getUiTheme = theme => {
  if (theme === undefined) return darkTheme;

  if (typeof theme === 'string') {
    if (theme === 'dark') {
      return darkTheme;
    } else if (theme === 'light') {
      return lightTheme;
    }

    throw Error(
      `Theme "${theme}" is not available. Please use "light" or "dark", or provide a custom theme object.`
    );
  }

  if (theme.base && theme.base === 'light') {
    return {
      ...lightTheme,
      ...theme
    };
  }

  return {
    ...darkTheme,
    ...theme
  };
};
