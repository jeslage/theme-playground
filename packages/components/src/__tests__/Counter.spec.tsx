import 'jest';
import * as React from 'react';
import { shallow } from 'enzyme';
import Counter from '../components/Counter/Counter';
import StyledCounter from '../components/Counter/Counter.style';
import { darkTheme } from '../theme';

describe('Counter', () => {
  it('Counter with decimal steps should render correctly', () => {
    const component = shallow(
      <Counter
        onChange={value => console.log(value)}
        label="Label"
        description="Testbeschreibung"
        value={1}
        steps={0.1}
        min={0}
        max={2}
        suffix="px"
      />
    );

    expect(component).toMatchSnapshot();
  });

  it('Counter with fixed steps should render correctly', () => {
    const component = shallow(
      <Counter
        onChange={value => console.log(value)}
        label="Label"
        description="Testbeschreibung"
        value={1}
        steps={1}
        min={0}
        max={2}
        suffix="px"
      />
    );

    expect(component).toMatchSnapshot();
  });

  it('StyledCounter should render correctly', () => {
    const component = shallow(<StyledCounter theme={{ colors: darkTheme }} />);

    expect(component).toMatchSnapshot();
  });
});
