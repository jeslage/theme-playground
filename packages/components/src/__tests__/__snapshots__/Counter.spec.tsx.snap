// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Counter Counter with decimal steps should render correctly 1`] = `
<styled.div>
  <Label
    description="Testbeschreibung"
    label="Label"
  />
  <div
    className="counter__counter"
  >
    <button
      aria-label="Decrease"
      disabled={false}
      onClick={[Function]}
      type="button"
    >
      <svg
        height="24"
        viewBox="0 0 24 24"
        width="24"
      >
        <rect
          height="2"
          rx="1"
          width="12"
          x="6"
          y="11"
        />
      </svg>
    </button>
    <span>
      <input
        onBlur={[Function]}
        onChange={[Function]}
        pattern="[0-9.]*"
        type="text"
        value={1}
      />
      px
    </span>
    <button
      aria-label="Increase"
      disabled={false}
      onClick={[Function]}
      type="button"
    >
      <svg
        height="24"
        viewBox="0 0 24 24"
        width="24"
      >
        <rect
          height="2"
          rx="1"
          width="12"
          x="6"
          y="11"
        />
        <rect
          height="12"
          rx="1"
          width="2"
          x="11"
          y="6"
        />
      </svg>
    </button>
  </div>
</styled.div>
`;

exports[`Counter Counter with fixed steps should render correctly 1`] = `
<styled.div>
  <Label
    description="Testbeschreibung"
    label="Label"
  />
  <div
    className="counter__counter"
  >
    <button
      aria-label="Decrease"
      disabled={false}
      onClick={[Function]}
      type="button"
    >
      <svg
        height="24"
        viewBox="0 0 24 24"
        width="24"
      >
        <rect
          height="2"
          rx="1"
          width="12"
          x="6"
          y="11"
        />
      </svg>
    </button>
    <span>
      <input
        onBlur={[Function]}
        onChange={[Function]}
        pattern="[0-9.]*"
        type="text"
        value={1}
      />
      px
    </span>
    <button
      aria-label="Increase"
      disabled={false}
      onClick={[Function]}
      type="button"
    >
      <svg
        height="24"
        viewBox="0 0 24 24"
        width="24"
      >
        <rect
          height="2"
          rx="1"
          width="12"
          x="6"
          y="11"
        />
        <rect
          height="12"
          rx="1"
          width="2"
          x="11"
          y="6"
        />
      </svg>
    </button>
  </div>
</styled.div>
`;

exports[`Counter StyledCounter should render correctly 1`] = `
<div
  className="sc-AxjAm hKdwgN"
/>
`;
